frameworks/autosklearn/exec.py

```diff
@@ -119,8 +136,28 @@ def save_artifacts(estimator, config):
             models_file = os.path.join(output_subdir('models', config), 'models.txt')
             with open(models_file, 'w') as f:
                 f.write(models_repr)
-    except Exception:
-        log.debug("Error when saving artifacts.", exc_info=True)
+        if 'debug_as_files' in artifacts or 'debug_as_zip' in artifacts:
+            print('Saving debug artifacts!')
+            debug_dir = output_subdir('debug', config)
+            ignore_extensions = ['.npy', '.pcs', '.model', '.ensemble', '.pkl']
+            tmp_directory = estimator.automl_._backend.temporary_directory
+            if 'debug_as_files' in artifacts:
+                files_to_copy = []
+                for r, d, f in os.walk(tmp_directory):
+                    for file_name in f:
+                        base, ext = os.path.splitext(file_name)
+                        if ext not in ignore_extensions:
+                            files_to_copy.append(os.path.join(r, file_name))
+                for filename in files_to_copy:
+                    dst = filename.replace(tmp_directory, debug_dir+'/')
+                    os.makedirs(os.path.dirname(dst), exist_ok=True)
+                    shutil.copyfile(filename, dst)
```

Comment:
FYI, there's also a `utils.walk_apply` function for this kind of logic similar to the `utils.zip_path`:
```
utils.walk_apply(directory, 
                           function with signature fn(path, is_dir), 
                           filtr=function with signature fn(path)
                           )
```
not sure this will simplify code much more here though, at least for the filtering logic, for the copying logic, the first callback would still use:
```
dst = filename.replace(tmp_directory, debug_dir+'/')
os.makedirs(os.path.dirname(dst), exist_ok=True)
shutil.copyfile(filename, dst)
```

Reply:
You mean like:
```python
                def _copy(filename):
                    dst = filename.replace(tmp_directory, debug_dir + '/')
                    os.makedirs(os.path.dirname(dst), exist_ok=True)
                    shutil.copyfile(filename, dst)
                utils.walk_apply(
                    tmp_directory,
                    lambda path, is_dir: _copy(path),
                    filtr=lambda path: (
                        os.path.splitext(path) not in ignore_extensions 
                        and not os.path.isdir(path)
                    ),  
                )
```?

It's a few lines shorter, but I'm not sure if it's more readable. What do you think?

Reply:
yeah, difference in length is minimal.
it's up to you, I tend to prefer high-order functions personally: I see better what's going on.

Reply:
btw, if you define 
```
def _copy(filename, **args):
```
then you don't need the lambda param, and it's more readable imo.

Reply:
Sorry, I can't follow. Could you please give a full example?

Reply:
```
                def _copy(filename, **_):
                    dst = filename.replace(tmp_directory, debug_dir + '/')
                    os.makedirs(os.path.dirname(dst), exist_ok=True)
                    shutil.copyfile(filename, dst)
                utils.walk_apply(
                    tmp_directory,
                    _copy,
                    filtr=lambda path: (
                        os.path.splitext(path) not in ignore_extensions 
                        and not os.path.isdir(path)
                    ),  
                )
```

Reply:
Done, thanks for providing me the exact code to do so.